beacon-chain/dbcleanup/service.go

```diff
@@ -0,0 +1,119 @@
+// Package dbcleanup defines the life cycle and logic of beacon DB cleanup routine.
+package dbcleanup
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/types"
+	"github.com/prysmaticlabs/prysm/shared/event"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+	"github.com/sirupsen/logrus"
+)
+
+var log = logrus.WithField("prefix", "dbcleaner")
+
+type chainService interface {
+	CanonicalCrystallizedStateFeed() *event.Feed
+}
+
+// CleanupService represents a service that handles routine task for cleaning up
+// beacon DB so our DB won't grow infinitely.
+// Currently it only cleans up block vote cache. In future, it could add more tasks
+// such as cleaning up historical beacon states.
+type CleanupService struct {
+	ctx                context.Context
+	cancel             context.CancelFunc
+	beaconDB           *db.BeaconDB
+	chainService       chainService
+	canonicalStateChan chan *types.CrystallizedState
+}
+
+// Config defines the needed fields for creating a new cleanup service.
+type Config struct {
+	SubscriptionBuf int
+	BeaconDB        *db.BeaconDB
+	ChainService    chainService
+}
+
+// NewCleanupService creates a new cleanup service instance.
+func NewCleanupService(ctx context.Context, cfg *Config) *CleanupService {
+	ctx, cancel := context.WithCancel(ctx)
+	return &CleanupService{
+		ctx:                ctx,
+		cancel:             cancel,
+		beaconDB:           cfg.BeaconDB,
+		chainService:       cfg.ChainService,
+		canonicalStateChan: make(chan *types.CrystallizedState, cfg.SubscriptionBuf),
+	}
+}
+
+// Start a cleanup service.
+func (d *CleanupService) Start() {
+	log.Infoln("Starting cleanup service")
+	go d.cleanDB()
+}
+
+// Stop a cleanup service.
+func (d *CleanupService) Stop() error {
+	defer d.cancel()
+
+	log.Info("Stopping cleanup service")
+	return nil
+}
+
+func (d *CleanupService) cleanDB() {
+	cStateSub := d.chainService.CanonicalCrystallizedStateFeed().Subscribe(d.canonicalStateChan)
+	defer cStateSub.Unsubscribe()
+
+	for {
+		select {
+		case <-d.ctx.Done():
+			log.Debug("Cleanup service context closed, exiting goroutine")
+			return
+		case cState := <-d.canonicalStateChan:
+			if err := d.cleanBlockVoteCache(cState.LastFinalizedSlot()); err != nil {
+				log.Errorf("Failed to clean block vote cache: %v", err)
+			}
+		}
+	}
+}
+
+func (d *CleanupService) cleanBlockVoteCache(latestFinalizedSlot uint64) error {
+	var lastCleanedFinalizedSlot uint64
+	var err error
+
+	lastCleanedFinalizedSlot, err = d.beaconDB.GetCleanedFinalizedSlot()
+	if err != nil {
+		log.Warnf("Cannot find last cleaned finalized slot, assume it to be 0")
+		lastCleanedFinalizedSlot = uint64(0)
+	}
+
+	log.Infof("Finalized slot: latest: %v, last cleaned: %v, %d blocks' vote cache will be cleaned",
```

Comment:
all the formatting should be in the form of `%d`

Reply:
Yeah you are right. Thanks!